### PR TITLE
Add linear point interpolation and extrapolation

### PR DESCRIPTION
**Pull request by znicholls**
Add support for point timeseries with linear interpolation as well as for constant extrapolation.

- [x] Implementation finished
- [x] Tests added
- [x] Documentation added
- [x] Example added (in the documentation, to an existing notebook, or in a new notebook)
- [x] Description in ``CHANGELOG.rst`` added (single line such as: ``(`#XX <https://github.com/openclimatedata/openscm/pull/XX>`_) Added feature which does something``)

### CODE REVIEW

File: openscm/timeseries_converter.py [hunk 1]
```diff
@@ -235,30 +233,53 @@ def _calc_continuous_representation(
             interpolated timeseries at that point in time ("y-value").
         """
         if self._timeseries_type == ParameterType.AVERAGE_TIMESERIES:
-            if self._interpolation_type == InterpolationType.LINEAR:
```
Comment (swillner) [inline openscm/timeseries_converter.py]: So we want to stick to these interpolation points regardless of the order of interpolation used?

File: openscm/timeseries_converter.py [hunk 2]
```diff
@@ -235,30 +233,53 @@ def _calc_continuous_representation(
...
+
+    def _get_scipy_extrapolation_args(self, values: np.ndarray) -> dict:
+        if self._extrapolation_type == ExtrapolationType.LINEAR:
+            return {"fill_value": "extrapolate"}
+        if self._extrapolation_type == ExtrapolationType.CONSTANT:
+            return {"fill_value": (values[0], values[-1]), "bounds_error": False}
+        # TODO: add cubic support
+        return {}
+
+    def _get_scipy_interpolation_arg(self):
```
Comment (swillner) [inline openscm/timeseries_converter.py]:
```suggestion
    def _get_scipy_interpolation_arg(self) -> str:
```

File: openscm/timeseries_converter.py [hunk 3]
```diff
@@ -235,30 +233,53 @@ def _calc_continuous_representation(
...
+                time_points,
+                values,
+                kind=self._get_scipy_interpolation_arg(),
+                **self._get_scipy_extrapolation_args(values),
+            )  # type: Callable[[float], float]
+            return res_point
+
+        raise NotImplementedError
+
+    def _get_scipy_extrapolation_args(self, values: np.ndarray) -> dict:
```
Comment (swillner) [inline openscm/timeseries_converter.py]:
```suggestion
    def _get_scipy_extrapolation_args(self, values: np.ndarray) -> dict[str, Any]:
```

Comment (swillner) [reply openscm/timeseries_converter.py]: Probably also need to import `Any` from `typing` at the beginning of the file.

File: openscm/timeseries_converter.py [hunk 4]
```diff
@@ -293,11 +314,33 @@ def _convert(
...
+            )
+            raise InsufficientDataError(error_msg)
+
+    def _convert_unsafe(
+        self,
+        values: np.ndarray,
+        source_time_points: np.ndarray,
+        target_time_points: np.ndarray,
+    ) -> np.ndarray:
+        """Same as ``self._convert`` but without error checking
```
Comment (swillner) [inline openscm/timeseries_converter.py]: Proper documentatin missing here (Just copy from `_convert`, which should probably “wrap _convert_safe while checking for...”)

File: notebooks/emissions-units-with-pint.ipynb [hunk 5]
```diff
@@ -908,15 +908,15 @@
...
-       " 'NOx_conversions': <pint.context.Context at 0x115633550>}"
+       "{'textile': <pint.context.Context at 0x118ba0390>,\n",
+       " 'spectroscopy': <pint.context.Context at 0x118bc37f0>,\n",
+       " 'sp': <pint.context.Context at 0x118bc37f0>,\n",
+       " 'boltzmann': <pint.context.Context at 0x118bc3780>,\n",
+       " 'chemistry': <pint.context.Context at 0x118bc3d30>,\n",
+       " 'chem': <pint.context.Context at 0x118bc3d30>,\n",
+       " 'AR4GWP12': <pint.context.Context at 0x118c42128>,\n",
+       " 'CH4_conversions': <pint.context.Context at 0x118bfc1d0>,\n",
+       " 'NOx_conversions': <pint.context.Context at 0x118c96400>}"
```
[the remark this answers is not in the document]
Comment (znicholls) [reply notebooks/emissions-units-with-pint.ipynb]: super suggestion done

File: notebooks/timeseries-conversion.ipynb [hunk 6]
```diff
@@ -106,13 +273,228 @@
...
+   "cell_type": "code",
+   "execution_count": 0,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stderr",
+     "output_type": "stream",
+     "text": [
+      "Traceback (most recent call last):\n",
+      "  File \"/Users/zebedeenicholls/Documents/AGCEC/MCastle/openscm/openscm/timeseries_converter.py\", line 319, in _convert\n",
```
[the remark this answers is not in the document]
Comment (znicholls) [reply notebooks/timeseries-conversion.ipynb]: I've had a go, see what you think

File: notebooks/timeseries-conversion.ipynb [hunk 7]
```diff
@@ -141,6 +523,65 @@
...
+   "execution_count": 0,
+   "metadata": {},
+   "outputs": [
+    {
+     "data": {
+      "text/plain": [
+       "array([ 0., 10., 20., 30., 40., 50., 60., 70., 80.])"
+      ]
+     },
+     "execution_count": 13,
```
[the remark this answers is not in the document]
Comment (znicholls) [reply notebooks/timeseries-conversion.ipynb]: the numbers or the execution count?

File: openscm/timeseries_converter.py (hunk 1 above)
Comment (znicholls) [reply openscm/timeseries_converter.py]: oops removing that was an errror good pick up!

File: notebooks/timeseries-conversion.ipynb (hunk 7 above)
[the remark this answers is not in the document]
Comment (swillner) [reply notebooks/timeseries-conversion.ipynb]: The execution count. Hmm, looks like they are part of the standard. Will update the cleanup in a separate PR.

File: notebooks/timeseries-conversion.ipynb [hunk 8]
```diff
@@ -106,13 +276,236 @@
...
+    "        (\"Emissions\", \"CH4\"),\n",
+    "        (\"World\",),\n",
+    "        \"MtCH4 / yr\",\n",
+    "        target_times,\n",
+    "        timeseries_type=ParameterType.POINT_TIMESERIES,\n",
+    "        interpolation_type=InterpolationType.LINEAR,\n",
+    "        extrapolation_type=extrap_type,\n",
+    "    )\n",
     "\n",
-    "If your data is periodic, all we need to do to specify our timeseries is to provide the start time and period length of the time a
```
Comment (swillner) [inline notebooks/timeseries-conversion.ipynb]:
```suggestion
    "           \"/\", \n",
```

Comment (swillner) [inline notebooks/timeseries-conversion.ipynb]:
```suggestion
    "    a = re.sub(\n",  # strip path from filename in traceback
```

Comment (znicholls) [reply notebooks/timeseries-conversion.ipynb]: ???

Comment (znicholls) [reply notebooks/timeseries-conversion.ipynb]: no i get it now my bad

Comment (swillner) [reply notebooks/timeseries-conversion.ipynb]: Oh, what happened here? Did I review on outdated source?

Comment (znicholls) [reply notebooks/timeseries-conversion.ipynb]: I think so...

**Comment by znicholls** [comment]
@swillner merge when ready if there's no more changes

**Pull request merged**